File: app/utils/dawproject/timeline/marker.ts
import { DoubleAdapter } from "../doubleAdapter";
import { Nameable } from "../nameable";
import type { IMarker } from "../types";
import { Utility } from "../utility";

export class Marker extends Nameable implements IMarker {
  time: number;

  constructor(time?: number, name?: string, color?: string, comment?: string) {
    super(name, color, comment);
    // Make time optional for deserialization, fromXmlObject will set it
    this.time = time || 0; // Provide a default placeholder
  }

  toXmlObject(): any {
    const attributes = super.toXmlObject(); // get attributes from Nameable

    // add required time attribute
    Utility.addAttribute(attributes, "time", this, {
      required: true,
      adapter: DoubleAdapter.toXml,
    });

    return attributes;
  }

  fromXmlObject(xmlObject: any): this {
    super.fromXmlObject(xmlObject); // populate inherited attributes from Nameable

    // validate and populate required time attribute
    Utility.populateAttribute<number>(xmlObject, "time", this, {
      required: true,
      adapter: DoubleAdapter.fromXml,
    });

    return this;
  }
}

File: app/utils/dawproject_python/marker.py
from classes.nameable import Nameable


class Marker(Nameable):
    def __init__(self, time, name=None, color=None, comment=None):
        super().__init__(name, color, comment)
        self.time = time

    def to_xml(self):
        marker_elem = super().to_xml()
        marker_elem.tag = "Marker"
        marker_elem.set("time", str(self.time))
        return marker_elem

    @classmethod
    def from_xml(cls, element):
        instance = super().from_xml(element)
        time = element.get("time")
        instance.time = float(time) if time is not None else None
        return instance
